Ed NC death handling

taltamir opened the pull request:
Ed NC death handling:

* boolean edAcquireHP()
** Cleaner handling of restoring HP as ed
** Prevents infinite loop when dying while mining in Itznotyerzitz Mine
** boolean edUnderworldAdv() created and used. If we are stuck at 0HP as ed and do not have a restore item, it will automatically spend 1 adv to "rest" by entering the underworld via the gate at his pyramid, shopping, then leaving.
* boolean ed_DelayNC_DailyDungeon() created and used. delays daily dungeon if we can die to DD traps and do not have a linen bandage to restore. As dying to the NC wastes an adv. No override allowed here since it is nonessential.
* boolean ed_DelayNC(int potential_dmg) created and used. Will return true if we are ed, can be killed by an NC dealing potential_dmg and lack the means to restore afterwards without wasting an adventure
* boolean ed_DelayNC(float potential_dmg_percent) created and used
** delays sonofa beach if we do not have linen bandages. as dying to zerg rush without a restore item wastes an adv.
* boolean L12_sonofaBeach() & L12_sonofaPrefix()
** use edAcquireHP() instead of specifically using linen bandages
** do not assume success

## How Has This Been Tested?

ash calls
playing ed on 4 accounts

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have based by pull request against the [beta branch](https://github.com/Loathing-Associates-Scripting-Society/autoscend/tree/beta) or have a good reason not to.

soolar reviewed (commented):
I feel like it might be worth the effort to determine how much damage you would take from these noncombats, and only skip them if that would actually kill you. Maybe not though, I forget how much effort that would be for these specific non-combats.

Malibu-Stacey commented on RELEASE/scripts/autoscend/auto_post_adv.ash:
```diff
@@ -1033,6 +1033,18 @@ boolean auto_post_adventure()
 		}
 	}
+	
+	if(isActuallyEd() && my_hp() == 0)
+	{
+		acquireHP(1);
+		if(my_hp() == 0 && get_property("_auto_edNoLinenCheck").to_boolean())	//we got beaten up in a risky NC as ed
+		{
+			if(!edUnderworldAdv())	//try to spend 1 adv to restore HP
+			{
+				abort("Failed to rest as ed");
+			}
+		}
+	}
```
We already have a call to `acquireHP(1)` for Ed in auto_pre_adv. Why is this necessary here rather than there?

taltamir replied:
I put it in post adv to catch _auto_edNoLinenCheck before it is reset, as well as to preempt our counter incrementing for dying in combat.
Thus making sure we are at 0 HP because of a noncombat that killed us rather than a combat loss.
The latter are being tracked separately to make sure we are not dying too many times

Maybe I should rework this.

taltamir replied:
change made

Malibu-Stacey reviewed (changes requested):
I think most of this is unnecessarily over complicated. The only thing needed is the `edUnderworldAdv()` function and updating the call to `acquireHP(1)` in auto_pre_adv to call it if we're still on 0 hp after acquireHP() returns.

taltamir commented:
> I feel like it might be worth the effort to determine how much damage you would take from these noncombats, and only skip them if that would actually kill you. Maybe not though, I forget how much effort that would be for these specific non-combats.

@soolar 

I just looked into the possibility of doing so.

Wiki says:
Sonofa beach deals between 50% to 100% of your max HP in damage. As such there is nothing to check, we should always seek to have curative before doing it to avoid wasting adventures.

Daily dungeon trap deals 50% of max HP. Daily dungeon door deals ???. If I knew how much door deals I could easily add a check for this. But wiki does not say

taltamir commented:
@Malibu-Stacey requested changes made

taltamir commented:
I hit an infinite loop in Itznotyerzitz Mine because it is not going through pre_adventure() and instead just calling acquireHP(1);
I realized that part of the issue is that we are over complicating ed restore by going through the full acquireHP function which is overly complex.

I split off ed healing into its own function `boolean edAcquireHP()`.
If ed and we call acquireHP as ed it will `return edAcquireHP()`
pre_adventure as ed will directly call it as ed instead of doing `acquireHP(1)`

It automatically knows to only restore ed if he is at 0 HP. And if we do not have restore items it will use edUnderworldAdv() to restore instead.

taltamir commented:
@soolar @Malibu-Stacey

I added some more code to solve some additional issues I came across.

In regards to your change requests, I tried to simplify it while also adding in the ability to specify how much HP we need to survive.

As I result I also moved the daily dungeon thing to its own function as it works differently and has some unique checks to do.

For sonofa beach I now have two versions:
1. I made a simplified version that just hardcodes the sonofa beach test
ed_DelayNC_SonofaBeach
commit bdaeb9c

2. I made a version that takes an estimated dmg and uses that to determine if we should delay or not
delayNC take estimated damage
commit 1bb73e8

Currently the overall PR is using number 2. but if i drop that commit it will revert to number 1.

Malibu-Stacey commented on RELEASE/scripts/autoscend/quests/level_12.ash:
```diff
@@ -1106,6 +1106,10 @@ boolean L12_sonofaBeach()
 	{
 		return false;
 	}
+	if(ed_DelayNC(100.0))	//zerg rush can deal 100% of maxHP in damage
+	{
+		return false;	//ed is not prepared. delay
+	}
```
This could infinitely prolong clearing the beach until there's nothing else left to do.
I don't think this is a good solution to this. At the point we're doing the level 12 quest Ka and thus linen bandages should be abundant and if not the cost of a single adventure to visit the underworld is better than stalling out the run indefinitely.

taltamir replied:
1. It specifically checks for linen bandages and does not delay if we have linen bandages.
2. It specifically checks if we are about to powerlevel so it does not stall indefinitely.
